Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,505 @@
+    static bool shouldPreferLinear = false;
+
+    std::call_once(onceFlag, [] {
+        if (const char* envString = getenv("WEBKIT_GPU_BUFFERS_PREFER_LINEAR_MODIFIER")) {
+            auto envStringView = StringView::fromLatin1(envString);
+            if (envStringView == "1"_s)
+                shouldPreferLinear = true;
+        }
+    });
+
+    return shouldPreferLinear;
+}
+
+std::unique_ptr<MemoryMappedGPUBuffer> MemoryMappedGPUBuffer::create(const IntSize& size, bool isOpaque)
+{
+    auto& manager = WebCore::DRMDeviceManager::singleton();
+    ASSERT(manager.isInitialized());
+
+    auto deviceNode = manager.mainDeviceNode(WebCore::DRMDeviceManager::NodeType::Primary);
+    if (deviceNode)
+        LOG(Images, "[MemoryMappedGPUBuffer] Allocating dma-bufs from device '%s'", deviceNode->filename().data());
+
+    auto* device = deviceNode->gbmDevice();
+    if (!device) {
+        LOG(Images, "[MemoryMappedGPUBuffer] Failed to get GBM device");
+        return nullptr;
+    }
+
+    const uint32_t preferredDMABufFormat = isOpaque ? DRM_FORMAT_XRGB8888 : DRM_FORMAT_ARGB8888;
+
+    const auto& supportedFormats = PlatformDisplay::sharedDisplay().dmabufFormats();
+    if (dumpSupportedDRMFormatsAndModifiers) {
+        LOG(Images, "[MemoryMappedGPUBuffer] Detected %lu supported dma-buf formats:", supportedFormats.size());
+        for (unsigned formatIndex = 0; formatIndex < supportedFormats.size(); ++formatIndex) {
+            const auto& format = supportedFormats[formatIndex];
+            LOG(Images, "format[%2u] -> fourcc=0x%2x, n_modifiers=%zu", formatIndex, format.fourcc, format.modifiers.size());
+
+            for (unsigned modifierIndex = 0; modifierIndex < format.modifiers.size(); ++modifierIndex)
+                LOG(Images, " -> modifier[%2u]=0x%02lx", modifierIndex, format.modifiers[modifierIndex]);
+        }
+    }
+
+    auto negotiateBufferFormat = [&]() -> std::optional<GLDisplay::DMABufFormat> {
+        GLDisplay::DMABufFormat useFormat;
+        for (const auto& format : supportedFormats) {
+            if (dumpBufferFormatNegotiation)
+                LOG(Images, "[MemoryMappedGPUBuffer] Testing buffer format for compatibility, fourcc=0x%2x, n_modifiers=%zu", format.fourcc, format.modifiers.size());
+
+            if (format.fourcc != preferredDMABufFormat) {
+                if (dumpBufferFormatNegotiation)
+                    LOG(Images, " -> incompatible format, does not match our preference, skipping.");
+                continue;
+            }
+
+            bool hasLinearModifier = false;
+            bool hasVivanteSuperTiledModifier = false;
+            if (useDRMModifiers) {
+                for (const auto& modifier : format.modifiers) {
+                    if (modifier == DRM_FORMAT_MOD_LINEAR)
+                        hasLinearModifier = true;
+                    else if (modifier == DRM_FORMAT_MOD_VIVANTE_SUPER_TILED)
+                        hasVivanteSuperTiledModifier = true;
```

Comment:
```suggestion
                    if (modifier == DRM_FORMAT_MOD_LINEAR)
                        hasLinearModifier = true;
                    else if (modifier == DRM_FORMAT_MOD_VIVANTE_SUPER_TILED)
                        hasVivanteSuperTiledModifier = true;
                    if (hasLinearModifier && hasVivanteSuperTiledModifier)
                        break
```

I understand we want to find out whether these two modifiers are supported. After we have done so, I think we can break. If we have a preference for either, then we could break earlier, but I didn't want to change the behaviour in my suggestion

Reply:
I simplified it, please re-check.